cleverhans/attacks.py

```diff
@@ -1,4 +1,342 @@
+    hard labels for this attack; no label smoothing.
+    Paper link: https://arxiv.org/pdf/1607.02533.pdf
+    """
+    def __init__(self, model, back='tf', sess=None, params={'eps': 0.05,
+                                                            'nb_iter': 10,
+                                                            'y': None,
+                                                            'ord': 'np.inf',
+                                                            'clip_min': None,
+                                                            'clip_max': None}):
+        """
+        Create a BasicIterativeMethod instance.
+
+        Attack-specific parameters:
+        :param eps: (required float) step size for each attack iteration
+        :param nb_iter: (required int) Number of attack iterations.
+        :param y: (required) A placeholder for the model labels.
+        :param ord: (optional) Order of the norm (mimics Numpy).
+                    Possible values: np.inf, 1 or 2.
+        :param clip_min: (optional float) Minimum input component value
+        :param clip_max: (optional float) Maximum input component value
+        """
+        super(BasicIterativeMethod, self).__init__(model, back, sess, params)
+
+        # Check that all required attack specific parameters are defined
+        req = ('eps', 'nb_iter', 'y')
+        if not all(k in params for k in req):
+            raise Exception("Attack requires label placeholder.")
+
+        # Save attack-specific parameters
+        self.eps = params['eps']
+        self.nb_iter = params['nb_iter']
+        self.y = params['y']
+        self.ord = params['ord'] if 'ord' in params else np.inf
+        self.clip_min = params['clip_min'] if 'clip_min' in params else None
+        self.clip_max = params['clip_max'] if 'clip_max' in params else None
+
+        # Check if order of the norm is acceptable given current implementation
+        if self.ord not in [np.inf, 1, 2]:
+            raise Exception("Norm order must be either np.inf, 1, or 2.")
+        if back == 'th':
+            error_string = "BasicIterativeMethod is not implemented in Theano"
+            raise NotImplementedError(error_string)
+
+    def generate(self, x, params={}):
+        import tensorflow as tf
+
+        # Initialize loop variables
+        adv_x = x
+        y = None
+        model_preds = self.model(adv_x)
+
+        for i in range(self.nb_iter):
+            FGSM = FastGradientMethod(self.model, back=self.back,
+                                      sess=self.sess,
+                                      params={'eps': self.eps,
+                                              'clip_min': self.clip_min,
+                                              'clip_max': self.clip_max,
+                                              'y': y})
+            adv_x = FGSM.generate(adv_x)
+            # After first iteration, we fix the labels to the first model
```

Comment:
This is using `model_preds` defined on the original `x`, so it's not really after the first iteration.
I'm not sure why you'd want to wait for the first iteration to finish though.
It seems like you should just do this before the loop starts.